src/filter/kuadrant_filter.rs

```diff
@@ -16,7 +16,8 @@ pub(crate) struct KuadrantFilter {
     header_resolver: Rc<HeaderResolver>,
 
     grpc_message_receiver_operation: Option<GrpcMessageReceiverOperation>,
-    headers_operations: Vec<HeadersOperation>,
+    response_headers_to_add: Option<Headers>,
+    request_headers_to_add: Option<Headers>,
```

Comment:
These have been renamed and are `Option` that is initialised once and extended until it is taken. Crucially if this is `None` it can no longer be initialised, we have added the headers and we are no longer in the correct phase